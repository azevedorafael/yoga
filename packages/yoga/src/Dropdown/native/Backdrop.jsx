import React, { useEffect, useState } from 'react';
import styled, { withTheme } from 'styled-components';
import { node, string, func, bool } from 'prop-types';
import {
  Animated,
  Dimensions,
  TouchableWithoutFeedback,
  Easing,
} from 'react-native';

import { Text } from '@gympass/yoga';

const { width, height } = Dimensions.get('screen');

const BackdropStyled = styled.View`
  position: absolute;
  z-index: 1;

  width: ${width};
  height: ${height};
  flex: 1;
`;

const ClosableArea = styled(Animated.View)`
  ${({
    theme: {
      yoga: { colors },
    },
  }) => `
    position: relative;
    width: 100%;
    height: 100%;

    background: ${colors.dark};
  `}
`;

const ContentWrapper = styled(Animated.View)`
  position: absolute;
  justify-content: flex-end;
`;

const Content = styled.View`
  ${({
    theme: {
      yoga: {
        components: { dropdown },
      },
    },
  }) => `
    justify-content: center;

    width: ${width};
    min-height: ${dropdown.backdrop.content.minHeight}px;

    background-color: ${dropdown.backdrop.content.backgroundColor};
    border-top-left-radius: ${dropdown.backdrop.content.border.radius.topLeft}px;
    border-top-right-radius: ${dropdown.backdrop.content.border.radius.topRight}px;
  `}
`;

const Title = styled(Text.H4)`
  width: 100%;
  padding: 20px 0;

  text-align: center;
`;

const Backdrop = ({
  visible,
  title,
  children,
  onClose,
  theme: {
    yoga: {
      transition,
      components: { dropdown },
    },
  },
  ...props
}) => {
  const [isRendered, setIsRendered] = useState(visible);
  const [backgroundAnimation] = useState(new Animated.Value(0));
  const [contentAnimation] = useState(
    new Animated.Value(-dropdown.backdrop.content.minHeight),
  );

  const animate = (animation, toValue) =>
    Animated.timing(animation, {
      toValue,
      duration: transition.duration[1],
      easing: Easing.out(Easing.ease),
    }).start(() => !visible && setIsRendered(false));

  useEffect(() => {
    if (visible) setIsRendered(true);
    animate(backgroundAnimation, visible ? 0.5 : 0);
    animate(
      contentAnimation,
      visible ? 70 : -dropdown.backdrop.content.minHeight,
    );
  }, [visible]);

  return (
    isRendered && (
      <BackdropStyled {...props}>
        <TouchableWithoutFeedback onPress={() => onClose()}>
          <ClosableArea style={{ opacity: backgroundAnimation }} />
        </TouchableWithoutFeedback>
        <ContentWrapper style={{ bottom: contentAnimation }}>
          <Content>
            {title && <Title>{title}</Title>}
            {children}
          </Content>
        </ContentWrapper>
      </BackdropStyled>
    )
  );
};

Backdrop.propTypes = {
  visible: bool,
  title: string,
  children: node.isRequired,
  onClose: func,
};

Backdrop.defaultProps = {
  visible: false,
  title: null,
  onClose: () => {},
};

export default withTheme(Backdrop);
