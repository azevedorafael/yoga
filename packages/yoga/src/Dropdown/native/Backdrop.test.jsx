import React from 'react';
import { render } from '@testing-library/react-native';
import { ThemeProvider } from '../..';
import Backdrop from './Backdrop';

jest.useFakeTimers();

describe('<Backdrop />', () => {
  describe('Snapshots', () => {
    it('should match snapshot', () => {
      const { container } = render(
        <ThemeProvider>
          <Backdrop title="Find an Activity">Some Content</Backdrop>
        </ThemeProvider>,
      );
      expect(container).toMatchSnapshot();
    });
    it('should match snapshot when visible', () => {
      const { container } = render(
        <ThemeProvider>
          <Backdrop visible title="Find an Activity">
            Some Content
          </Backdrop>
        </ThemeProvider>,
      );
      expect(container).toMatchSnapshot();
    });
  });
});
