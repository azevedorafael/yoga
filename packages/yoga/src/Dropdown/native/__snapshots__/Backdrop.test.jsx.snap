// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Backdrop /> Snapshots should match snapshot 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`<Backdrop /> Snapshots should match snapshot when visible 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Modal
    animationType="none"
    hardwareAccelerated={false}
    transparent={true}
    visible={true}
  >
    <TouchableWithoutFeedback>
      <View
        style={
          Object {
            "backgroundColor": "#41414A",
            "height": "100%",
            "opacity": 0,
            "position": "relative",
            "width": "100%",
          }
        }
      />
    </TouchableWithoutFeedback>
    <View
      style={
        Object {
          "bottom": -320,
          "justifyContent": "flex-end",
          "position": "absolute",
          "width": "100%",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#FFFFFF",
              "borderTopLeftRadius": 8,
              "borderTopRightRadius": 8,
              "justifyContent": "center",
              "width": "100%",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#41414A",
                "fontFamily": "Open Sans",
                "fontSize": 16,
                "fontWeight": "700",
                "lineHeight": 20,
                "marginBottom": 0,
                "marginLeft": 0,
                "marginRight": 0,
                "marginTop": 0,
                "paddingBottom": 20,
                "paddingLeft": 0,
                "paddingRight": 0,
                "paddingTop": 20,
                "textAlign": "center",
                "width": "100%",
              },
            ]
          }
        >
          Find an Activity
        </Text>
        Some Content
      </View>
    </View>
  </Modal>
</View>
`;
